src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+           ID_statement_expression),
+      "quantifier must not contain side effects");
+
+    // Handle the case that quantifier body is a statement expression.
+    if(
+      code.operands()[0].id() == ID_side_effect &&
+      code.operands()[0].get_named_sub()[ID_statement].id() ==
+        ID_statement_expression)
+    {
+      goto_programt where;
+      goto_convert(code, where, mode);
+      where.compute_location_numbers();
+
+      natural_loops_mutablet natural_loops(where);
+      INVARIANT(
+        natural_loops.loop_map.size() == 0,
+        "quantifier must not contain side effects");
+
+      // `last` is the instruction corresponding to the last expression in the
+      // statement expression.
+      goto_programt::const_targett last;
+      for(goto_programt::const_targett it = where.instructions.begin();
+          it != where.instructions.end();
+          ++it)
+      {
+        // `last` is an other-instruction.
+        if(it->is_other())
+        {
+          last = it;
+        }
+      }
+      auto last_expr =
+        expr_try_dynamic_cast<code_expressiont>(last->get_other())
+          ->expression();
+
+      // `pathes` contains all the `targett` we are iterating over.
+      std::vector<goto_programt::const_targett> pathes;
+      pathes.push_back(where.instructions.begin());
+      std::vector<std::pair<exprt, replace_symbolt>>
+        path_conditions_and_value_maps;
+      path_conditions_and_value_maps.push_back(
+        std::make_pair(true_exprt(), replace_symbolt()));
+
+      std::unordered_set<symbol_exprt, irep_hash> declared_symbols;
+      // All bounded variables are local.
+      declared_symbols.insert(qex.variables().begin(), qex.variables().end());
+
+      exprt res = true_exprt();
+
+      // Visit the quantifeir body along `pathes`.
+      while(!pathes.empty())
+      {
+        auto &current_it = pathes.back();
+        auto &path_condition = path_conditions_and_value_maps.back().first;
+        auto &value_map = path_conditions_and_value_maps.back().second;
+        INVARIANT(
+          current_it != where.instructions.end(),
+          "Quantifier body must have a unique end expression.");
+
+        // Add all local-declared symbols into `declared_symbols`.
+        if(current_it->is_decl())
+        {
+          declared_symbols.insert(current_it->decl_symbol());
+        }
+        // ASSIGN lhs := rhs
+        // Add the replace lhr <- value_map(rhs) to the current value_map.
+        if(current_it->is_assign())
+        {
+          // Check that if lhs is a declared symbol.
+          auto lhs = current_it->assign_lhs();
+          INVARIANT(
+            declared_symbols.count(find_base_symbol(lhs)),
+            "quantifier must not contain side effects");
+          exprt rhs = current_it->assign_rhs();
+          INVARIANT(
+            !has_subexpr(rhs, ID_function_call),
+            "quantifier must not contain side effects");
```

Comment:
As `rhs` has been subject to all rules of goto conversion (and its cleaning) there really shouldn't ben any kind of side effect left on the right-hand side of an assignment. I believe this `INVARIANT` is quite possibly pointless as `ID_function_call` expressions can't possibly exist at that point.